### ISSUE TITLE
Updating items in an optionally nested list breaks typing

### ISSUE THREAD

**Issue by AndersBennedsgaard** [microsoft/pyright#5479]
**Describe the bug**
I wanted to update a list of `None`s and other lists using a for-loop (data coming from a big JSON blob), which fails since the type of the current item is not inferred in the primary list.

**Code or Screenshots**
My explanation above is probably quite confusing, so here is a code example instead:

```python
lst: list[list[str] | str | None] = [["a", "b"], None, ["c"], "test"]
for i, item in enumerate(lst):
    if item is None:
        continue
    if isinstance(item, str):
        continue

    # reveal_type(item) reports 'Type of "item" is "list[str]"'
    if item[0] == "a":
        # below breaks with Pylance reporting both
        #   "__setitem__" method not defined on type "str"
        #   Object of type "None" is not subscriptable
        lst[i][0] = "something else"
```

It is possible to set `item[0] = "something else"` and `lst[i] = item`, but this gets annoying for deeply nested lists, so it could be nice to have the type inferred back to the primary list somehow (I don't know if this is even possible)

**Comment by erictraut** [comment]
Pyright is doing the right thing here from a type checking perspective. The type of expression `lst[i]` is `list[str] | str | None`, and you're attempting to index this value. If the element is a `str` or `None`, the operation will fail. Not surprisingly, mypy emits the same error on the same line.

Your logic is relying on specialized knowledge about the relationship between `item` and `lst[i]` — knowledge that a static type checker doesn't know about. You would need to modify your logic if you want it to type check.

For example, you can do the following. This works because the type of `item` has been narrowed to `list[str]`.

```python
    if item[0] == "a":
        item[0] = "something else"
```
